data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -3,300 +3,282 @@
+        } else if (csb.getTokenProvider() != null) {
             tokenProvider = csb.getTokenProvider();
+            authenticationType = AuthenticationType.AAD_ACCESS_TOKEN_PROVIDER;
         } else {
-            authenticationType = AuthenticationType.AAD_DEVICE_LOGIN;
+            if (StringUtils.isNotBlank(csb.getUserUsername())) {
+                userUsername = csb.getUserUsername();
+            }
+            redirectUri = new URI("http://localhost");
+            authenticationType = AuthenticationType.AAD_USER_PROMPT;
         }
 
-        // Set the AAD Authority URI
-        String aadAuthorityId = (csb.getAuthorityId() == null ? DEFAULT_AAD_TENANT : csb.getAuthorityId());
+        aadAuthorityUrl = determineAadAuthorityUrl(csb.getAuthorityId());
+    }
+
+    private String determineAadAuthorityUrl(String aadAuthorityUrl) {
+        String aadAuthorityId = (aadAuthorityUrl == null ? ORGANIZATION_URI_SUFFIX : aadAuthorityUrl);
         String aadAuthorityFromEnv = System.getenv("AadAuthorityUri");
-        if (aadAuthorityFromEnv == null){
-            aadAuthorityUri = String.format("https://login.microsoftonline.com/%s", aadAuthorityId);
+        if (aadAuthorityFromEnv == null) {
+            aadAuthorityUrl = String.format("https://login.microsoftonline.com/%s", aadAuthorityId);
         } else {
-            aadAuthorityUri = String.format("%s%s%s", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", aadAuthorityId);
+            aadAuthorityUrl = String.format("%s%s%s", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", aadAuthorityId);
         }
+        return aadAuthorityUrl;
     }
 
-    String acquireAccessToken() throws DataServiceException, DataClientException {
+    protected String acquireAccessToken() throws DataServiceException, DataClientException {
         if (authenticationType == AuthenticationType.AAD_ACCESS_TOKEN) {
             return accessToken;
         }
 
         if (authenticationType == AuthenticationType.AAD_ACCESS_TOKEN_PROVIDER) {
             try {
-                return tokenProvider.call();
+                Future<String> future = executor.submit(tokenProvider);
+                return getAsyncResultNonblocking(future);
             } catch (Exception e) {
-                throw new DataClientException(clusterUrl, e.getMessage(), e);
+                throw new DataClientException(scope, e.getMessage(), e);
             }
         }
 
-        if (lastAuthenticationResult == null) {
-            acquireToken();
-        } else if (IsInvalidToken()) {
-            if (lastAuthenticationResult.getRefreshToken() == null) {
-                acquireToken();
-            } else {
-                lastAuthenticationResultLock.lock();
+        try {
+            lastClientApplicationLock.lock();
+            if (lastAuthenticationResult == null) {
+                lastAuthenticationResult = acquireNewAccessToken();
+            } else if (isInvalidToken()) {
```

Comment:
In a sense, we don't need to check this because MSAL's acquireTokenSilently() already checks whether the token is valid. However, the behavior we've had in the Java SDK until now was to consider the token invalid even if it'll only expire a minute from now, so I'm maintaining that same functionality.